packages/regenerator-transform/src/visit.js

```diff
@@ -177,32 +184,118 @@ function getOuterFnExpr(funPath) {
     node.id = funPath.scope.parent.generateUidIdentifier("callee");
   }
 
+  let outerFnExpr = node.id
+
   if (node.generator && // Non-generator functions don't need to be marked.
-      t.isFunctionDeclaration(node)) {
-    let pp = funPath.findParent(function (path) {
+    t.isFunctionDeclaration(node)) {
+    let pp = funPath.findParent(function(path) {
       return path.isProgram() || path.isBlockStatement();
     });
 
     if (!pp) {
-      return node.id;
+      return funPath;
     }
+    const parameters = funPath.node.params.map(function(param) {
+      return t.cloneDeep(param);
+    })
+
+    // turn our FunctionDeclaration into a FunctionExpression with a different
+    // name - our our eventual outerFnExpr needs to be identical to the original
+    // FunctionDeclaration name so that external calls work correctly, and we
+    // can't also use that name in our FunctionExpression because it would shadow
+    // the outerFnExpr in our wrapping scope
+    const functionExpression = t.functionExpression(
+      funPath.scope.generateUidIdentifierBasedOnNode(node.id),
+      parameters,
+      t.cloneDeep(funPath.node.body),
+      funPath.node.generator,
+      funPath.node.async
+    );
 
-    let markDecl = getRuntimeMarkDecl(pp);
-    let markedArray = markDecl.declarations[0].id;
-    let funDeclIdArray = markDecl.declarations[0].init.callee.object;
-    t.assertArrayExpression(funDeclIdArray);
+    // in spite of the above, we want the resulting function's name property
+    // property to be the same as the original so we wrap the FunctionExpression
+    // with:
+    //   Object.defineProperty(functionExpression, 'name', { value: outerFnExpr.name }
+    // this way introspection / prototype chains work the way we'd expect
```

Comment:
i made the change locally - all tests pass except [this one](https://github.com/facebook/regenerator/blob/master/test/tests.es6.js#L1579):
```
  1) function declaration hoisting should not interfere with function rebinding:
     AssertionError [ERR_ASSERTION]: [GeneratorFunction: toBeRebound] deepEqual 42
      at check (test/shared.js:16:12)
      at Context.<anonymous> (test/tests.es5.js:3307:5)
```

i _think_ it's because of the shadowing issue i mentioned. the original transform generates this code:
```
var _marked = [toBeRebound].map(regeneratorRuntime.mark);

function rebindTo(value) {
  var oldValue = toBeRebound;
  toBeRebound = value;
  return oldValue;
}

function toBeRebound() {
  var originalValue;
  return regeneratorRuntime.wrap(function toBeRebound$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          originalValue = toBeRebound;
          _context.next = 3;
          return toBeRebound;

        case 3:
          assert.strictEqual(rebindTo(42), originalValue);
          _context.next = 6;
          return toBeRebound;

        case 6:
          assert.strictEqual(rebindTo("asdf"), 42);
          _context.next = 9;
          return toBeRebound;

        case 9:
        case "end":
          return _context.stop();
      }
    }
  }, _marked[0], this);
}
```

and my (newly modified) code generates this code:
```
var toBeRebound = /*#__PURE__*/regeneratorRuntime.mark(function toBeRebound() {
  var originalValue;
  return regeneratorRuntime.wrap(function toBeRebound$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          originalValue = toBeRebound;
          _context.next = 3;
          return toBeRebound;

        case 3:
          assert.strictEqual(rebindTo(42), originalValue);
          _context.next = 6;
          return toBeRebound;

        case 6:
          assert.strictEqual(rebindTo("asdf"), 42);
          _context.next = 9;
          return toBeRebound;

        case 9:
        case "end":
          return _context.stop();
      }
    }
  }, toBeRebound, this);
});
function rebindTo(value) {
  var oldValue = toBeRebound;
  toBeRebound = value;
  return oldValue;
}
```
i think that `toBeRebound = value` call inside of `rebindTo` is changing `toBeRebound` in the main scope, but `function toBeRebound()` is still shadowing that value inside of the generator.